src/main/java/bio/terra/common/retry/SamRetry.java

```diff
@@ -0,0 +1,188 @@
+    private static final int MAXIMUM_WAIT_SECONDS = 30;
+    private static final int INITIAL_WAIT_SECONDS = 10;
+    private static final int OPERATION_TIMEOUT_SECONDS = 300;
+    private final Instant operationTimeout;
+    private int retrySeconds;
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    SamRetry() {
+        this.operationTimeout = now().plusSeconds(OPERATION_TIMEOUT_SECONDS);
+        this.retrySeconds = INITIAL_WAIT_SECONDS;
+    }
+
+    @VisibleForTesting
+    SamRetry(int timeoutSeconds) {
+        this.operationTimeout = now().plusSeconds(timeoutSeconds);
+        this.retrySeconds = INITIAL_WAIT_SECONDS;
+    }
+
+    @FunctionalInterface
+    public interface SamVoidFunction {
+        void apply() throws ApiException, InterruptedException;
+    }
+
+    @FunctionalInterface
+    public interface SamFunction<R> {
+        R apply() throws ApiException, InterruptedException;
+    }
+
+
+    static <T> T retry(SamFunction<T> function)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry();
+        return samRetry.perform(function);
+    }
+
+    // Version of the retry function for testing so that we can override timeout seconds
+    @VisibleForTesting
+    static <T> T retry(SamFunction<T> function, int timeoutSeconds)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry(timeoutSeconds);
+        return samRetry.perform(function);
+    }
+
+    static void retry(SamVoidFunction function)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry();
+        samRetry.performVoid(function);
+    }
+
+    // Version of the retry function for testing so that we can override timeout seconds
+    @VisibleForTesting
+    static void retry(SamVoidFunction function, int timeoutSeconds)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry(timeoutSeconds);
+        samRetry.performVoid(function);
+    }
+
+    private <T> T perform(SamFunction<T> function) throws InterruptedException {
+        while (true) {
+            try {
+                return function.apply();
+            } catch (ApiException ex) {
+                handleApiException(ex);
+            } catch (Exception ex) {
+                throw new InternalServerErrorException("Unexpected exception type: " + ex.toString(), ex);
+            }
+            sleepBeforeRetrying();
+        }
+    }
+
+    private void performVoid(SamVoidFunction function) throws InterruptedException {
+        while (true) {
+            try {
+                function.apply();
+                return;
+            } catch (ApiException ex) {
+                handleApiException(ex);
+            } catch (Exception ex) {
+                throw new InternalServerErrorException("Unexpected exception type: " + ex.toString(), ex);
+            }
+            sleepBeforeRetrying();
+        }
+    }
+
+    private void handleApiException(ApiException ex) {
```

Comment:
Hmmm, I'm not sure that we want to force users of the API to have their exceptions wrapped. Maybe offer two versions of retry functions, or offer the exception wrapping as a separate composable function? Not sure.

Reply:
I think we should force a mapping based on exceptions from the common library that can be easily converted to an ErrorReport with the right HTTP status by a global exception handler.

At a minimum, we should convert it to an exception that has a message. The API messages from Sam are hard to use. Sometimes you have to dig into the next cause level to get any message text. There is code in TDR that does that.

Reply:
Per offline convo with Will, going to leave these wrapped in the common exception classes. Adding more Sam specific exceptions did not seem to add much at this point